Support PHP 5.2

taggon opened the pull request:
See #5.

As WordPress supports PHP 5.2, we're supposed to do so. This PR will downgrade and modify the PHP scripts to support the old version.

Todo:
- Server scripts
  - [x] Component_Themes
  - [x] Component_Themes_Api
  - [x] Component_Themes_Api_wrapper
  - [x] Component_Themes_Builder
  - [x] Component_Themes_Component
  - [x] Component_Themes_Styles
  - [x] React
  - [x] helper functions
- Components
  - [x] ColumnComponent
  - [x] FooterText
  - [x] HeaderText
  - [x] MenuWidget
  - [x] PageLayout
  - [x] PostAuthor
  - [x] PostBody
  - [x] PostContent
  - [x] PostDate
  - [x] PostList
  - [x] PostTitle
  - [x] RowComponent
  - [x] SearchWidget
  - [x] SinglePost
  - [x] TextWidget

taggon committed:
Support PHP 5.2

As Wordpress supports PHP 5.2, we're supposed to do so.
This commit is a starting point for supporting PHP 5.2 and includes following changes:
- `Component_Themes_Api::api_data_wrapper` returns an instance of `Component_Themes_Api_Wrapper` instead of a closure.
- Array shorthand syntax `[]` are replaced with `array()`.
- If there's no registered component of the type, `Component_Themes_Builder::get_component_by_type` returns an instance of `Component_Themes_Not_Found_Component`.
- `PostList` component works with PHP 5.2 now.

taggon committed:
Downgrade some components and React class.

As now the following components work with PHP 5.2.

- ColumnComponent
- HeaderText
- MenuWidget
- PostBody
- PostContent
- PostTitle
- RowComponent

taggon committed:
Remove array shorthand syntax from ColumnComponent class.

taggon committed:
Remove null rendered children when  renders

taggon committed:
Remove array shorthand syntax from TextWidget component.

taggon committed:
Downgrade SearchWidget and SinglePost

by replacing shorthand array syntax with `array()` and using named functions instead of closures.

taggon committed:
Remove styled component closure.

...because PHP 5.2 doesn't support closures.
Instead, you can style components using `$styles` static property.

For more detail, see the documentation.

taggon committed:
Replace shorthand array syntax with `array()`.

taggon committed:
Remove shorthand array syntax

taggon committed:
Remove shorthand array syntax

taggon committed:
Downgrade the Builder class

taggon committed:
Downgrade the Styles class.

taggon committed:
Remove yarn.lock from the repo.

taggon committed:
Remove shorthand array syntax.

sirbrillig commented on server/class-component-themes-api-wrapper.php:
```diff
@@ -0,0 +1,31 @@
+		$api_props = ct_get_value( $context, 'apiProps', array() );
+
+		// merge api_props to the context
+		foreach ( $api_props as $name=>$prop ) {
```
(Style) Can you add spaces around the `=>`?

sirbrillig commented on server/class-component-themes-builder.php:
```diff
@@ -28,21 +28,25 @@ public function render() {
-			if ( ! is_string( $prop ) ) {
-				return '';
+		$attrs = array();
+		foreach( $this->props as $name=>$value ) {
```
(Style) Can you add spaces around `=>`?

sirbrillig commented on server/class-component-themes-builder.php:
```diff
@@ -98,7 +106,8 @@ public function create_element( $component, $props = [], $children = [] ) {
 		if ( 'Component_Themes_Stateless_Component' === $component ) {
 			// This would be an error state, but to prevent errors, we will make a noop
-			return new Component_Themes_Stateless_Component( function() {}, $props, $children );
+			$noop = create_function( '', '' );
```
TIL! 👍

sirbrillig commented on server/class-component-themes-builder.php:
```diff
@@ -170,11 +200,11 @@ private function expand_config_partials( $component_config, $partials ) {
-			$children = array_map( $expander, $component_config['children'] );
-			$new_config = array_merge( $component_config, [ 'children' => $children ] );
+			$chilren = array();
+			foreach( $component_config['children'] as $key=>$child ) {
```
(Style) Spaces around `=>`

sirbrillig commented on server/class-component-themes-builder.php:
```diff
@@ -170,11 +200,11 @@ private function expand_config_partials( $component_config, $partials ) {
-			};
-			$children = array_map( $expander, $component_config['children'] );
-			$new_config = array_merge( $component_config, [ 'children' => $children ] );
+			$chilren = array();
```
s/$chilren/$children/

sirbrillig commented on server/class-component-themes-styles.php:
```diff
@@ -66,7 +64,7 @@ private function prepend_namespace_to_node( $namespace, $val ) {
 
 	public function prepend_namespace_to_style_string( $namespace, $styles ) {
-		$parser = new \CssParser();
+		$parser = new CssParser();
```
Whoops! Nice find.

sirbrillig commented on server/class-component-themes-component-wrapper.php:
```diff
@@ -0,0 +1,21 @@
+			return $component->get_styles();
+		}
+
+		if ( is_string( $component) && is_subclass_of( $component, 'Component_Themes_Component' ) ) {
```
(Style) There's a missing space after `is_string( $component`

sirbrillig commented on server/class-react.php:
```diff
@@ -27,29 +29,25 @@ public static function render( $component ) {
-			return ( $child );
-		} );
+		$rendered_children = React::mapChildren( $children, 'React::render' );
+		$rendered_children = array_filter( $rendered_children, 'ct_not_empty' );
```
Nice!

sirbrillig commented on src/themes/components/MenuWidget/index.php:
```diff
@@ -1,39 +1,45 @@
+
 	public function render() {
-		$title = $this->get_prop( 'title' );
+		$title = $this->get_prop( 'title', '' );
```
Setting a default here will cause the title markup to be shown even if the title is empty because the next line uses `isset()`. The intent was to not show the title area at all if no title is present.

sirbrillig commented on src/themes/components/SinglePost/index.php:
```diff
@@ -1,30 +1,33 @@
-		return React::cloneElement( $child, $post_data );
-	} );
-	return React::createElement( 'div', [ 'className' => $class_name ], $new_children );
+	$new_children = React::cloneChileren( $children, $post_data );
```
s/cloneChileren/cloneChildren/

sirbrillig reviewed (changes requested):
This looks amazing. Nice work all around! A few things:

- There were a couple of typos I saw where I left a comment.
- You might want to install `phpcs` and the WordPress style pack to check your styles. I just installed it recently and learned a few things that I had been doing incorrectly. I marked a few places where I saw an error appear in my editor.
- When I try to run this branch on my test site, I get a bunch of PHP warnings. You can test the output by commenting-out line 85 of public/class-component-themes-public.php which will cause only the PHP version to be rendered.

taggon committed:
Fix typos and match coding standard.

taggon commented:
@sirbrillig Hey! I fixed all typos and styles. :)

sirbrillig committed:
Add ssr query string to disable client side rendering

Add `?ssr=true` to the URL to disable the javascript rendering. This
makes it easier to test server side rendering.

sirbrillig commented:
I'm still seeing errors:

```
Notice: Undefined variable: component in /srv/www/wordpress-default/wp-content/plugins/component-themes/server/class-component-themes-component-wrapper.php on line 11

Notice: Undefined variable: component in /srv/www/wordpress-default/wp-content/plugins/component-themes/server/class-component-themes-component-wrapper.php on line 15

Notice: Undefined variable: component in /srv/www/wordpress-default/wp-content/plugins/component-themes/server/class-component-themes-component-wrapper.php on line 11

Notice: Undefined variable: component in /srv/www/wordpress-default/wp-content/plugins/component-themes/server/class-component-themes-component-wrapper.php on line 15

Notice: Undefined variable: component in /srv/www/wordpress-default/wp-content/plugins/component-themes/server/class-component-themes-component-wrapper.php on line 11

Notice: Undefined variable: component in /srv/www/wordpress-default/wp-content/plugins/component-themes/server/class-component-themes-component-wrapper.php on line 15

Warning: array_merge(): Argument #1 is not an array in /srv/www/wordpress-default/wp-content/plugins/component-themes/server/class-component-themes-builder.php on line 85

Warning: Invalid argument supplied for foreach() in /srv/www/wordpress-default/wp-content/plugins/component-themes/server/class-component-themes-builder.php on line 38
```

I just pushed a commit to allow rendering a page using just the PHP. If you add `?ssr=true` to the URL, it will not activate the javascript rendering. This should make it easier to test.

taggon committed:
Merge branch 'master' into support/php-5.2

taggon committed:
Fix references to undefined variables.

- In PHP 5.2, component `$styles` static property should be public because parent class' static method cannot access child class' non-public static property.

- Builder create_element() can be called with `null` props argument. When the argument is null, create_element() changes it into an empty array before array merging it.

taggon commented:
@sirbrillig I wasn't able to see the errors because `WP_DEBUG` was turned off in my local machine. My bad. 😛  Now all the errors seem to be fixed.

sirbrillig commented:
Nice! Errors are gone. 

Testing the `home` template works great. The `post` template, however, seems to not be getting the data from the API and I'm not sure why (I'm just seeing the default "No title", "No date", etc.).

To test this template, click on the post title of any post listed on the home page.

taggon committed:
Fix: global api data is not passed to Single Post component.

taggon committed:
Fix: global api data is not passed to Single Post component.

sirbrillig commented:
Nice work!
